src/libpgagroal/ev.c

```diff
@@ -0,0 +1,1667 @@
+static int __io_uring_receive_handler(struct ev_loop* ev, struct ev_io* w, struct io_uring_cqe* cqe,
+                                      void** unused, int* bid, bool is_proxy);
+static int __io_uring_send_handler(struct ev_loop* ev, struct ev_io* w, struct io_uring_cqe* cqe);
+static int __io_uring_accept_handler(struct ev_loop* ev, struct ev_io* w, struct io_uring_cqe* cqe);
+static int __io_uring_periodic_handler(struct ev_loop* ev, struct ev_periodic* w);
+#endif
+
+#if HAVE_EPOLL
+static int __epoll_init(struct ev_loop*);
+static int __epoll_destroy(struct ev_loop*);
+static int __epoll_handler(struct ev_loop*, void*);
+static int __epoll_loop(struct ev_loop*);
+static int __epoll_fork(struct ev_loop**);
+static int __epoll_io_start(struct ev_loop*, struct ev_io*);
+static int __epoll_io_stop(struct ev_loop*, struct ev_io*);
+static int __epoll_io_handler(struct ev_loop*, struct ev_io*);
+static int __epoll_send_handler(struct ev_loop*, struct ev_io*);
+static int __epoll_accept_handler(struct ev_loop*, struct ev_io*);
+static int __epoll_receive_handler(struct ev_loop*, struct ev_io*);
+static int __epoll_periodic_init(struct ev_periodic*, int);
+static int __epoll_periodic_start(struct ev_loop*, struct ev_periodic*);
+static int __epoll_periodic_stop(struct ev_loop*, struct ev_periodic*);
+static int __epoll_periodic_handler(struct ev_loop*, struct ev_periodic*);
+static int __epoll_signal_stop(struct ev_loop*, struct ev_signal*);
+static int __epoll_signal_handler(struct ev_loop*);
+static int __epoll_signal_start(struct ev_loop*, struct ev_signal*);
+inline static int
+__epoll_set_non_blocking(int fd)
+{
+   int flags = fcntl(fd, F_GETFL, 0);
+   if (flags == -1)
+   {
+      return EV_ERROR;
+   }
+   return fcntl(fd, F_SETFL, flags | O_NONBLOCK);
+}
+#endif
+
+static inline bool
+__is_running(struct ev_loop* ev)
+{
+   return ev->running;
+}
+static inline bool
+__is_running_atomic(struct ev_loop* ev)
+{
+   return atomic_load(&ev->atomic_running);
+}
+
+static inline void
+__set_running(struct ev_loop* ev)
+{
+   ev->running = true;
+}
+static inline void
+__set_running_atomic(struct ev_loop* ev)
+{
+   atomic_store(&ev->atomic_running, true);
+}
+
+static inline void
+__break(struct ev_loop* loop)
+{
+   loop->running = false;
+}
+static inline void
+__break_atomic(struct ev_loop* loop)
+{
+   atomic_store(&loop->atomic_running, false);
+}
+
+static enum ev_backend backend_value(char* backend);
+
+void
+pgagroal_ev_print_backends(void)
+{
+   int cnt = 0;
+   char log[MISC_LENGTH] = { 0 };
+#if HAVE_URING
+   strcat(log, "io_uring, ");
+   cnt++;
+#endif
+#if HAVE_EPOLL
+   strcat(log, "epoll, ");
+   cnt++;
+#endif
+#if HAVE_KQUEUE
+   strcat(log, "kqueue, ");
+   cnt++;
+#endif
+   if (cnt > 0)
+   {
+      log[strlen(log) - 2] = '\0';
+      pgagroal_log_debug("available ev backends: %s", log);
+   }
+   else
+   {
+      pgagroal_log_debug("no ev backends available");
```

Comment:
I think this should be at an higher level, like `warn`.

Reply:
If there is no supported EV backend it should be FATAL and result in exit(1)